pinot-common/src/test/java/com/linkedin/pinot/common/partition/PartitionAssignmentGeneratorTest.java

```diff
@@ -0,0 +1,352 @@
+
+    // 3 partitions 2 replicas, 2 on 0th seq number 1 on 1st seq number (8 segments)
+    instances = idealStateBuilder.getInstances(0, 0);
+    idealState =
+        idealStateBuilder.transitionToState(0, 0, "ONLINE").addConsumingSegment(0, 1, numReplicas, instances).build();
+    verifyPartitionAssignmentFromIdealState(tableConfig, idealState, numPartitions);
+
+    // 3 partitions 2 replicas, all on 1st seg num (12 segments)
+    instances = idealStateBuilder.getInstances(1, 0);
+    idealStateBuilder.transitionToState(1, 0, "ONLINE").addConsumingSegment(1, 1, numReplicas, instances);
+    instances = idealStateBuilder.getInstances(2, 0);
+    idealState =
+        idealStateBuilder.transitionToState(2, 0, "ONLINE").addConsumingSegment(2, 1, numReplicas, instances).build();
+    verifyPartitionAssignmentFromIdealState(tableConfig, idealState, numPartitions);
+
+    // 3 partitions 2 replicas, seq 0 has moved to COMPLETED servers, seq 1 on CONSUMING instances
+    instances = Lists.newArrayList("s1_completed", "s2_completed");
+    idealState = idealStateBuilder.moveToServers(0, 0, instances).build();
+    verifyPartitionAssignmentFromIdealState(tableConfig, idealState, numPartitions);
+
+    // status of latest segments OFFLINE/ERROR
+    idealState = idealStateBuilder.transitionToState(1, 1, "ERROR").transitionToState(2, 1, "OFFLINE").build();
+    verifyPartitionAssignmentFromIdealState(tableConfig, idealState, numPartitions);
+
+    // all non llc segments
+    String aNonLLCSegment = "randomName";
+    Map<String, String> instanceStateMap = new HashMap<>(1);
+    instanceStateMap.put("s1", "CONSUMING");
+    instanceStateMap.put("s2", "CONSUMING");
+    idealState = idealStateBuilder.clear().addSegment(aNonLLCSegment, instanceStateMap).build();
+    numPartitions = 0;
+    verifyPartitionAssignmentFromIdealState(tableConfig, idealState, numPartitions);
+
+    // some non llc segments, some llc segments
+    numPartitions = 1;
+    instances = getConsumingInstanceList(6);
+    idealState = idealStateBuilder.addConsumingSegments(numPartitions, 0, numReplicas, instances).build();
+    verifyPartitionAssignmentFromIdealState(tableConfig, idealState, numPartitions);
+  }
+
+  private void verifyPartitionAssignmentFromIdealState(TableConfig tableConfig, IdealState idealState,
+      int numPartitions) {
+    PartitionAssignment partitionAssignmentFromIdealState =
+        _partitionAssignmentGenerator.getPartitionAssignmentFromIdealState(tableConfig, idealState);
+    Assert.assertEquals(tableConfig.getTableName(), partitionAssignmentFromIdealState.getTableName());
+    Assert.assertEquals(partitionAssignmentFromIdealState.getNumPartitions(), numPartitions);
+    // check that latest segments are honoring partition assignment
+    Map<Integer, LLCSegmentName> partitionIdToLatestLLCSegment = new HashMap<>(numPartitions);
+    for (Map.Entry<String, Map<String, String>> entry : idealState.getRecord().getMapFields().entrySet()) {
```

Comment:
makes sense to add a method getLatestSegmentsAndPartitions() in the code, and then use that in the test as well?

Reply:
added